### PR TITLE
log in - log out

### PR DESCRIPTION
**Pull request by truongnv0211**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class SessionsController < ApplicationController
+
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove line empty

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  def new
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Reomve ( )

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 class UsersController < ApplicationController
+  def show
+    @user = User.find(params[:id])
```
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+module SessionsHelper
+
+  def log_in(user)
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    if (user_id = session[:user_id])
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      @current_user ||= User.find_by(id: session[:user_id])
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove  ( )

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    elsif (user_id = cookies.signed[:user_id])
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      user = User.find_by(id: user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove  ( )

File: app/helpers/sessions_helper.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      end
+    end
+  end
+
+  def logged_in?
+    !current_user.nil?
+  end
+
+  def log_out
+    session.delete(:user_id)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    @current_user = nil
+  end
+
+  def redirect_back_or(default)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/models/user.rb [hunk 10]
```diff
@@ -14,4 +14,19 @@ class User < ActiveRecord::Base
             length: {minimum: Settings.password_minimum}
 
   has_secure_password
+
+  def self.digest(string)
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove ( )

File: app/views/layouts/_header.html.erb [hunk 11]
```diff
@@ -5,7 +5,23 @@
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
-        <li><%= link_to t("log_in"), "#" %></li>
+       <% if logged_in? %>
+          <li><%= link_to "Users", users_path %></li>
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Account <b class="caret"></b>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: I18n

File: app/views/sessions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, t("log_in") %>
+<h1>Log in</h1>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: I18n

File: app/views/sessions/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: Remove ()

File: app/views/sessions/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: Break line

File: app/views/sessions/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
```
Comment (phamkykhoi) [inline app/views/sessions/new.html.erb]: Break line

File: app/views/layouts/_header.html.erb [hunk 16]
```diff
@@ -5,7 +5,23 @@
...
+            </a>
+            <ul class="dropdown-menu">
+              <li><%= link_to t("profile"), current_user %></li>
+              <li class="divider"></li>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: What is this tag li ?

File: app/views/layouts/_header.html.erb [hunk 17]
```diff
@@ -5,7 +5,23 @@
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
-        <li><%= link_to t("log_in"), "#" %></li>
+       <% if logged_in? %>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: Indent

File: app/helpers/sessions_helper.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def log_out
+    session.delete :user_id
+    @current_user = nil
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
+    session.delete(:forwarding_url)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/helpers/sessions_helper.rb [hunk 19]
```diff
@@ -0,0 +1,32 @@
+module SessionsHelper
+
+  def log_in(user)
```
Comment (phamkykhoi) [inline app/helpers/sessions_helper.rb]: Remove ( )

File: app/views/sessions/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,16 @@
+<% provide :title, t("log_in") %>
+<h1><%= t(:log_in) %></h1>
```
Comment (khanhkhanh) [inline app/views/sessions/new.html.erb]: Remove ()

File: app/controllers/sessions_controller.rb [hunk 21]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
```
Comment (khanhkhanh) [inline app/controllers/sessions_controller.rb]: Try removing "downcase" feature into model as you have learnt in tutorial :)

File: app/views/layouts/_header.html.erb [hunk 22]
```diff
@@ -5,7 +5,22 @@
       <ul class="nav navbar-nav navbar-right">
         <li><%= link_to t("home"), root_path %></li>
         <li><%= link_to t("help"), help_path %></li>
-        <li><%= link_to t("log_in"), "#" %></li>
+        <% if logged_in? %>
+          <li><%= link_to "Users", users_path %></li>
```
Comment (khanhkhanh) [inline app/views/layouts/_header.html.erb]: i18n

File: app/views/sessions/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    <%= form_for :session, url: login_path do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.submit t("log_in"), class: "btn btn-primary" %>
+    <% end %>
+    <p>New user? <%= link_to t("sign_up"), signup_path %></p>
```
Comment (khanhkhanh) [inline app/views/sessions/new.html.erb]: i18n

File: app/controllers/sessions_controller.rb (hunk 21 above)
Comment (truongnv0211) [inline app/controllers/sessions_controller.rb]: Session only have controller, doesn't have model

Comment (truongnv0211) [inline app/controllers/sessions_controller.rb]: In tutorial, when u log in case sensitive

File: app/controllers/sessions_controller.rb [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+    if user && user.authenticate(params[:session][:password])
```
Comment (phamkykhoi) [inline app/controllers/sessions_controller.rb]: Remove ( )

File: db/seeds.rb [hunk 25]
```diff
@@ -5,3 +6,55 @@
...
+  answer_1 = vietnameses[4*index]
+  answer_2 = vietnameses[4*index+1]
+  answer_3 = vietnameses[4*index+2]
+  answer_4 = vietnameses[4*index+3]
+  correct_answer = rand(1..4)
+  category_id = rand(1..5)
+  Word.create!(jp_word: jp_word,correct_answer: correct_answer,
+               answer_1: answer_1,answer_2: answer_2,
+               answer_3: answer_3,answer_4: answer_4,category_id: category_id)
+end
```
Comment (phamkykhoi) [inline db/seeds.rb]: This file, change => to :

File: app/controllers/sessions_controller.rb (hunk 21 above)
Comment (khanhkhanh) [inline app/controllers/sessions_controller.rb]: Sorry, I didn't look at these line carefully. I thought it's UsersController :)

Comment (truongnv0211) [inline app/controllers/sessions_controller.rb]: No problem, thank for your review

File: app/models/user.rb [hunk 26]
```diff
@@ -14,4 +14,19 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def self.new_token
+    SeureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (phamkykhoi) [inline app/models/user.rb]: update_attribute => update_attributes

File: app/helpers/sessions_helper.rb [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+    !current_user.nil?
+  end
+
+  def log_out
+    session.delete :user_id
+    @current_user = nil
+  end
+
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
```
Comment (ducminhvu) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/models/user.rb [hunk 28]
```diff
@@ -14,4 +14,19 @@ class User < ActiveRecord::Base
             length: {minimum: Settings.password_minimum}
 
   has_secure_password
+
+  def self.digest string
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                  BCrypt::Engine.cost
+    BCrypt::Password.create(string, cost: cost)
```
Comment (ducminhvu) [inline app/models/user.rb]: remove ()

File: app/models/user.rb [hunk 29]
```diff
@@ -14,4 +14,19 @@ class User < ActiveRecord::Base
...
+  end
+
+  def self.new_token
+    SeureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes(:remember_digest, User.digest(remember_token))
```
Comment (ducminhvu) [inline app/models/user.rb]: remove ()

File: app/controllers/sessions_controller.rb (hunk 24 above)
Comment (truongnv0211) [inline app/controllers/sessions_controller.rb]: When i remove it, my app can't run :(

File: app/models/user.rb [hunk 30]
```diff
@@ -14,4 +14,19 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create string, cost: cost
+  end
+
+  def self.new_token
+    SeureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes :remember_digest, User.digest(remember_token)
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove ( )

Comment (truongnv0211) [inline app/models/user.rb]: I try do it but it doesn't works

**Comment by dieunb** [comment]
OK

**Pull request merged**